### PR TITLE
[MainMenu] Adds Animated Background & DisplayMode Settings.

### LINKED ISSUE #815: [Main Menu] Animated Background & DisplayMode Settings.
### **Is your feature request related to a problem? Please describe.**
Not related to a problem, this is **feature** & **enchancement** request that is applicable to any game out there.

### **Describe the solution you'd like**
I know that right now bug fixing its the priority, but i'd like to add the following **feature** & **enchancements** to the base engine:
* Ability to set an **Animated** Background to the game's main menu.
* Ability to to set the display mode of fullscreen textures (stretch, centered, fitH, fitW, etc.).

### **Additional context**
Many games/mmos out there have animated intros and login screens, presentation its always important when it comes to any book/movie/game/form of art in general.

### PR DESCRIPTION
**Pull request by Arufonsu**
**Preview of Animated Background (DisplayMode: Stretch)**

https://user-images.githubusercontent.com/17498701/149450094-8ee3c759-cba1-4d5a-aee6-b7149f977f93.mp4

This Pull Request should resolve #815

**11-01-2022 ~ 12-01-2022**
* Adds the ability to to stretch the main menu's static background to the game client's window size though a new boolean property at Client's Config.json.
* Adds the ability to set an animated background in the main menu. Frames and their speed (Milliseconds) are configurable though new properties at Client's Config.json.
* Created the method "DrawFullScreenAnimatedTextureStretched", which draws a full screen animation as background, stretching it's size to the game client's window size.
* Created the method "DrawFullScreenAnimatedTexture", which draws a full screen animation as background, without stretching it's size to the game client's window size.
* Added an extra property to the client config that allows to toggle between stretched or not full screen animated background.
* [ClientConfig TODOs] Resolved a TODO with the property 'IntroImages', commented the property 'UpdateUrl' and did few minor improvements to some comments.
* As suggested by Cheshire92, drawing from a single giant sheet isn't very performance wise, so instead we will now load and animate from multiple frame files inside the animation resources folder. (ie: "main_menu_0.png, main_menu_1.png, main_menu_2.png, etc").
* The string property AnimatedMenuBgTexture doesn't needs to be set with the file the extension suffix (".png").

**13-01-2022**
* Undid unnecessary comments and TODOs.
* Removed unnecessary and redundant code.
* Renamed 'DrawFullScreenAnimatedTexture' to 'DrawFullScreenTextureAnimated' (in order to maintain consistency).
* 'DrawFullScreenTextureAnimated' is now capable to count the amount of frames required by animations. (no manual setup is required anymore)
* Created an Enum for 'DisplayModes'.
* Created the Method 'DrawFullScreenTextureCentered'.
* The Client is now capable to display full screen textures and animations in 5 different display modes.

[**Checkout the required resources to test this PR!**](https://github.com/AscensionGameDev/Intersect-Assets/pull/4)

### CODE REVIEW

File: Intersect.Client/Core/Graphics.cs [hunk 1]
```diff
@@ -144,12 +148,71 @@ public static void DrawIntro()
...
+                if (animTex == null)
+                {
+                    return;
+                }
+
+                var destRectangle = new FloatRect();
+                var textFrames = ClientConfiguration.Instance.AnimatedMenuBgFrameCount;
+                var animDelay = ClientConfiguration.Instance.AnimatedMenuBgDelay;
+
+                mainmenuBgAnimTime += 1;
```
Comment (AlexVild) [inline Intersect.Client/Core/Graphics.cs:167]: Does the background image change in speed when you change frame rate settings in the client settings menu? Curious since I don't see any sort of "Delta Time" game clock calculations. Also, doing the frame advancement in a way that's locked to game time will make it so that a user actually knows the unit of time they're entering into `AnimatedMenuBgDelay`.

This is just a problem I observed when I was adding my screen shake effect. For an example of a system that uses the game clock to calculate when to advance a transition, you can look at `Fade.cs` - that's what I copied off of lol

**Review by AlexVild — CHANGES_REQUESTED** [review]
Seems cool, and I'd certainly use it! But I think a tweak is necessary for this to function right. Could be wrong though.

File: Intersect.Client/Core/Graphics.cs (hunk 1 above)
Comment (Arufonsu) [reply Intersect.Client/Core/Graphics.cs:167]: will take a look at it!

**Comment by Cheshire92** [comment]
Is this loading off of a single giant sheet of frames?
That sounds incredibly inefficient on older hardware and even somewhat on modern hardware given the texture size.

Wouldn't it make more sense to have a list of frames to go through from a video memory perspective?

**Comment by Arufonsu** [comment]
> Is this loading off of a single giant sheet of frames? That sounds incredibly inefficient on older hardware and even somewhat on modern hardware given the texture size.
> 
> Wouldn't it make more sense to have a list of frames to go through from a video memory perspective?

Didn't really considered it an issue, since i never ran on any sort of issue here on my end to be honest (i don't have any sort of 'old hardware' to test this, so...)

**Comment by Arufonsu** [comment]
> Is this loading off of a single giant sheet of frames? That sounds incredibly inefficient on older hardware and even somewhat on modern hardware given the texture size.
> 
> Wouldn't it make more sense to have a list of frames to go through from a video memory perspective?

Didn't really considered it an issue, since i never ran on any sort of issue here on my end to be honest (i don't have any sort of 'old hardware' to test this, so...) and this is basically how everything is currently being animated all over around this engine.

_Screenshot: Performance of the game client while using an animated background in the main menu:_
imagen

_Screenshot: Performance of the game client while using an static and stretched background in the main menu:_
imagen

the performance seems to be very relative, if anything, in both cases, either the CPU and RAM barely jump past 7% CPU usage and 99 ~ 110 MB of RAM. I currently use an intel CPU (i3 9th GEN, not very 'new' to this day already).

If you happen to have an approach on how to do this with individual splitted frames, that would be amazing if it happens to be more performance friendly than this. Also yeah, if you use a sprite sheet that's too large, things are gonna get funky at some point, i'd say that this is working fine with sheets no bigger than 25k Pixels-wide ish ~

**Comment by Arufonsu** [comment]
> Is this loading off of a single giant sheet of frames? That sounds incredibly inefficient on older hardware and even somewhat on modern hardware given the texture size.
> 
> Wouldn't it make more sense to have a list of frames to go through from a video memory perspective?

Hi, thanks for the suggestion, i just did something about it so instead of loading from a single sheet, individual files are used as frames for the background animation. Let me know what you guys think now! 
[
Oh, you may want to test this PR with these updated resources as well](https://github.com/AscensionGameDev/Intersect-Assets/pull/4), had to force push both PRs in order to get things in order and properly working.

File: Intersect.Client/Core/Graphics.cs (hunk 1 above)
Comment (Arufonsu) [reply Intersect.Client/Core/Graphics.cs]: Hey, thanks for pointing this out, the timing for the animation frames should be working properly in milliseconds now ;)

File: Intersect.Client/Core/Graphics.cs [hunk 2]
```diff
@@ -768,6 +812,76 @@ public static void DrawFullScreenTextureFitMaximum(GameTexture tex)
...
+            var destRectangle = new FloatRect();
+            var srcRectangle = new FloatRect(0, 0, animTex.Width, animTex.Height);
+            destRectangle.X = 0;
+            destRectangle.Y = 0;
+            destRectangle.Width = srcRectangle.Width;
+            destRectangle.Height = srcRectangle.Height;
+            DrawGameTexture(animTex, srcRectangle, destRectangle, new Color(255, 255, 255, 255));
+        }
+
+        private static void DrawFullScreenAnimatedTextureStretched(
```
Comment (lodicolo) [inline Intersect.Client/Core/Graphics.cs:850]: These two methods are basically the same except for the destination rectangle, the code should get reused.

File: Intersect.Client/Core/Graphics.cs [hunk 3]
```diff
@@ -768,6 +812,76 @@ public static void DrawFullScreenTextureFitMaximum(GameTexture tex)
...
+            );
+
+            if (animTex == null)
+            {
+                return;
+            }
+
+            var destRectangle = new FloatRect();
+            var srcRectangle = new FloatRect(0, 0, animTex.Width, animTex.Height);
+            destRectangle.X = 0;
```
Comment (lodicolo) [inline Intersect.Client/Core/Graphics.cs:843]: Honestly the default behavior should probably be centering on the screen so you don't have an 800x600 animated background just sitting up in the top left corner of a 1080p screen

File: Intersect (Core)/Configuration/ClientConfiguration.cs [hunk 4]
```diff
@@ -69,58 +81,98 @@ public void Validate()
...
+        /// Toggles the ability to stretch the animated background in the main menu.
+        /// </summary>
+        public bool AnimatedMenuBgStretched { get; set; } = DEFAULT_ANIM_MENUBG_STRETCH;
 
+        /// <summary>
+        /// Sets the base name for the files to be used as frames in the animated background of the main menu.
+        /// ie: if you set "main_menu", you will need to place your animation frames inside the animation folder
+        /// and name them like this:  "main_menu_0.png, main_menu_1.png, main_menu_2.png, etc".
+        /// </summary>
+        public string AnimatedMenuBgTexture { get; set; } = DEFAULT_ANIM_MENUBG_TEXTURE;
```
Comment (lodicolo) [inline Intersect (Core)/Configuration/ClientConfiguration.cs:159]: Why not just reuse MenuBackground?

If the animated background feature is turned on, MenuBackground is the prefix, otherwise it's the exact name.

File: Intersect (Core)/Configuration/ClientConfiguration.cs [hunk 5]
```diff
@@ -69,58 +81,98 @@ public void Validate()
...
+        /// <summary>
+        /// Toggles the ability to stretch the static background in the main menu.
+        /// </summary>
+        public bool MenuBackgroundStretched { get; set; } = DEFAULT_MENUBG_STRETCH;
+
+        /// <summary>
+        /// Toggles the ability to draw an animated background in the main menu.
+        /// When enabled, the static background in the main menu won't be drawn.
+        /// </summary>
+        public bool AnimatedMenuBgEnabled { get; set; } = DEFAULT_ANIM_MENUBG_TOGGLE;
```
Comment (lodicolo) [inline Intersect (Core)/Configuration/ClientConfiguration.cs:147]: Background should be spelled out.

File: Intersect (Core)/Configuration/ClientConfiguration.cs [hunk 6]
```diff
@@ -69,58 +81,98 @@ public void Validate()
...
-        /// Menu music file name
+        /// Sets a music file to be played in the main menu.
         /// </summary>
         public string MenuMusic { get; set; } = DEFAULT_MENU_MUSIC;
 
         /// <summary>
-        /// Menu background art
+        /// Sets a list from which introductory images will be drawn when the game client starts.
+        /// </summary>
+        public List<string> IntroImages { get; set; } = new List<string>();
```
Comment (lodicolo) [inline Intersect (Core)/Configuration/ClientConfiguration.cs:131]: None of these documentation changes belong in this PR, they should be a separate PR.

File: Intersect (Core)/Configuration/ClientConfiguration.cs [hunk 7]
```diff
@@ -69,58 +81,98 @@ public void Validate()
...
 
         /// <summary>
-        /// Hostname of the server to connect to
+        /// Hostname of the server that the client will connect to.
         /// </summary>
         public string Host { get; set; } = DEFAULT_HOST;
 
         /// <summary>
-        /// Port of the server to connect to
+        /// Port of the server that the client will connect to.
```
Comment (lodicolo) [inline Intersect (Core)/Configuration/ClientConfiguration.cs:89]: I don't think the wording changes are needed here.

File: Intersect (Core)/Configuration/ClientConfiguration.cs [hunk 8]
```diff
@@ -69,58 +81,98 @@ public void Validate()
...
+        /// <summary>
+        /// Toggles the ability to draw an animated background in the main menu.
+        /// When enabled, the static background in the main menu won't be drawn.
+        /// </summary>
+        public bool AnimatedMenuBgEnabled { get; set; } = DEFAULT_ANIM_MENUBG_TOGGLE;
+
+        /// <summary>
+        /// Toggles the ability to stretch the animated background in the main menu.
+        /// </summary>
+        public bool AnimatedMenuBgStretched { get; set; } = DEFAULT_ANIM_MENUBG_STRETCH;
```
Comment (lodicolo) [inline Intersect (Core)/Configuration/ClientConfiguration.cs:152]: There only needs to be one menu background display-mode property.

File: Intersect (Core)/Configuration/ClientConfiguration.cs [hunk 9]
```diff
@@ -69,58 +81,98 @@ public void Validate()
...
+        /// ie: if you set "main_menu", you will need to place your animation frames inside the animation folder
+        /// and name them like this:  "main_menu_0.png, main_menu_1.png, main_menu_2.png, etc".
+        /// </summary>
+        public string AnimatedMenuBgTexture { get; set; } = DEFAULT_ANIM_MENUBG_TEXTURE;
+
+        /// <summary>
+        /// Sets the frame count of the animated background in the main menu.
+        /// Make sure to set this value depending on the amount of files to be used from the animation resources.
+        /// </summary>
+        public int AnimatedMenuBgFrameCount { get; set; } = DEFAULT_ANIM_MENUBG_FRAMECOUNT;
```
Comment (lodicolo) [inline Intersect (Core)/Configuration/ClientConfiguration.cs:165]: I feel like this shouldn't be necessary and enables sloppy asset management.

If there are 10 images, all 10 should be used. If only 5 of them are being used, then the extra 5 files don't belong, they're just wasting disk space. This should be automatically calculated in my opinion.

File: Intersect (Core)/Configuration/ClientConfiguration.cs [hunk 10]
```diff
@@ -69,58 +81,98 @@ public void Validate()
...
+
+        /// <summary>
+        /// Sets the frame speed (milliseconds) of the animated background in the main menu.
+        /// </summary>
+        public long AnimatedMenuBgFrameSpeed { get; set; } = DEFAULT_ANIM_MENUBG_FRAMESPEED;
```
Comment (lodicolo) [inline Intersect (Core)/Configuration/ClientConfiguration.cs:170]: Should be called `MenuBackgroundFrameInterval`. Millisecond is an interval, not a rate of change (which speed is).

File: Intersect (Core)/Configuration/ClientConfiguration.cs [hunk 11]
```diff
@@ -69,58 +81,98 @@ public void Validate()
...
+
+        /// <summary>
+        /// Optional - built in updater system.
+        /// Sets an URL to Package Updates (ie: https://freemmorpgmaker.com/updater/update.json).
+        /// </summary>
```
Comment (lodicolo) [inline Intersect (Core)/Configuration/ClientConfiguration.cs:175]: The optional part of this comment really isn't necessary.

This documentation change should not be a part of this PR though.

File: Intersect (Core)/Configuration/ClientConfiguration.cs [hunk 12]
```diff
@@ -69,58 +81,98 @@ public void Validate()
...
+        /// Sets a texture to be used as static background in the main menu.
         /// </summary>
         public string MenuBackground { get; set; } = DEFAULT_MENU_BACKGROUND;
 
-        // TODO: What is this for?
-        public List<string> IntroImages { get; set; } = new List<string>();
+        /// <summary>
+        /// Toggles the ability to stretch the static background in the main menu.
+        /// </summary>
+        public bool MenuBackgroundStretched { get; set; } = DEFAULT_MENUBG_STRETCH;
```
Comment (lodicolo) [inline Intersect (Core)/Configuration/ClientConfiguration.cs:141]: No. No. No. No. No.

This should be an enum of multiple display modes. If you only want to write the implementation for the default (center screen), and stretched, that's fine -- but it absolutely should not be a boolean.

These are all display modes that would be ideal to support, boolean is not good enough (and honestly stretching if the aspect ratio is different is hideous)

- Center screen
- Center horizontally
- Center vertically
- Fit (letter/pillarboxed)
- Cover (the smaller dimension fits, the larger one hangs off the edge of the screen)
- Stretched
- Tiled

**Review by lodicolo — CHANGES_REQUESTED** [review]

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 11 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs:175]: alright, i will port these changes afterwards then!

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 8 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs:152]: do you mean like, a single property that defines whether the (static or animated background) should be stretched?

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 9 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs:165]: yeah... i previously had to do it this way since i was using a sprite-sheet instead, i will change this logic for sure, thanks for pointing this out ~

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 12 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs:141]: decided to add the following for now:
imagen

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 10 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs:170]: Done 👍🏽

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 9 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs:165]: done 👍🏽

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 8 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs:152]: Done 👍🏽 also, created an Enum for "DisplayModes"

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 5 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs:147]: Done ~ 👍🏽

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 4 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs:159]: Done 👍🏽

File: Intersect.Client/Core/Graphics.cs (hunk 3 above)
Comment (Arufonsu) [reply Intersect.Client/Core/Graphics.cs:843]: realized that i was writing redundant code, thanks for pointing this out, i will see what i can do with a method to simply center the texture on screen as well.

File: Intersect.Client/Core/Graphics.cs (hunk 2 above)
Comment (Arufonsu) [reply Intersect.Client/Core/Graphics.cs:850]: realized that i was writing redundant code, thanks for pointing this out

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 7 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs:89]: undid these changes 👍🏽

File: Intersect (Core)/Configuration/ClientConfiguration.cs (hunk 12 above)
Comment (Arufonsu) [reply Intersect (Core)/Configuration/ClientConfiguration.cs]: Created the following Enum for now:

imagen

**Comment by Arufonsu** [comment]
After cleaning up a bunch of unnecessary stuff, this is what the client configuration file looks like now:
imagen

Test with Animated Background (DisplayMode: Center)
imagen

Test with Animated Background (DisplayMode: Stretch)
imagen

Test with Static Background (Random texture: Trade Window) (DisplayMode: Center)
imagen

Test with Static Background (Random texture: Trade Window) (DisplayMode: Stretch)
imagen

File: Intersect.Client/Core/Graphics.cs [hunk 13]
```diff
@@ -768,6 +827,60 @@ public static void DrawFullScreenTextureFitMaximum(GameTexture tex)
...
+
+                    break;
+                case DisplayModes.FitHeight:
+                    DrawFullScreenTextureFitHeight(animTex);
+
+                    break;
+                case DisplayModes.FitWidth:
+                    DrawFullScreenTextureFitWidth(animTex);
+
+                    break;
```
Comment (lodicolo) [inline Intersect.Client/Core/Graphics.cs:880]: Can you put the extra newline after the break, not before it

File: Intersect.Client/Core/Graphics.cs [hunk 14]
```diff
@@ -768,6 +827,60 @@ public static void DrawFullScreenTextureFitMaximum(GameTexture tex)
...
+
+            if (animTex == null)
+            {
+                return;
+            }
+
+            switch (displayMode)
+            {
+                case DisplayModes.Default:
+                    DrawFullScreenTexture(animTex);
```
Comment (lodicolo) [inline Intersect.Client/Core/Graphics.cs:862]: Is it just me or does this method (`DrawFullScreenTexture`, it's already part of the code base) look like it's doing the exact same thing as "stretch" does, but... in extra steps?

@Arufonsu @Cheshire92 @jcsnider

**Review by lodicolo — CHANGES_REQUESTED** [review]

**Comment by lodicolo** [comment]
Optional relatively simple additions @Arufonsu 

Add "Fit" (just the word fit) that uses the "FitMaximum" draw method?

Add "Cover" (just the word cover) that uses the "FitMinimum" draw method?

File: Intersect.Client/Core/Graphics.cs (hunk 14 above)
Comment (Arufonsu) [reply Intersect.Client/Core/Graphics.cs:862]: somehow it does and it doesn't?
i realized that if the texture itself is bigger than the game screen size, the 'default' DrawFullScreenTexture doesn't stretches down the texture, but it **does** stretches them up when the screen is bigger, so yeah they are similar but not exactly the same?

File: Intersect.Client/Core/Graphics.cs (hunk 13 above)
Comment (Arufonsu) [reply Intersect.Client/Core/Graphics.cs:880]: got it ~

**Comment by Arufonsu** [comment]
[Today's commit, as discussed in AGD's Discord Group.](https://github.com/AscensionGameDev/Intersect-Engine/pull/1044/commits/7950babd58ace1cfe46ae4088b06fe0d17ea0cdc)

https://user-images.githubusercontent.com/17498701/149581447-dbdd0d48-25bb-4d36-bc2d-0788c782b617.mp4

**Comment by Arufonsu** [comment]
04-03-2022: Signed up [the assets commit](https://github.com/AscensionGameDev/Intersect-Assets/pull/4)

**Review by AlexVild — DISMISSED** [review]
Pretty cool feature - code looks good, and if Panda's happy with the implementation then I'm certainly happy

File: Intersect.Client/Core/Graphics.cs [hunk 15]
```diff
@@ -151,51 +151,74 @@ public static void DrawIntro()
...
+                sMenuBackground = sContentManager.GetTexture(
+                    TextureType.Gui, ClientConfiguration.Instance.MenuBackground[sMenuBackgroundIndex]
                 );
+
+                if (sMenuBackground == null)
+                {
+                    return;
+                }
+
+                if (sMenuBackgroundInterval < Timing.Global.Milliseconds)
```
Comment (lodicolo) [inline Intersect.Client/Core/Graphics.cs:166]: Can you cache `Timing.Global.Milliseconds` in a variable here instead of recalling it on L169?

File: Intersect.Client/Core/Graphics.cs [hunk 16]
```diff
@@ -151,51 +151,74 @@ public static void DrawIntro()
...
+                {
+                    sMenuBackgroundIndex++;
+                    sMenuBackgroundInterval = Timing.Global.Milliseconds +
+                                              ClientConfiguration.Instance.MenuBackgroundFrameInterval;
+                }
+
+                if (sMenuBackgroundIndex >= ClientConfiguration.Instance.MenuBackground.Count)
+                {
+                    sMenuBackgroundIndex = 0;
```
Comment (lodicolo) [inline Intersect.Client/Core/Graphics.cs:175]: I think `sMenuBackgroundIndex = (sMenuBackgroundIndex + 1) % ClientConfiguration.Instance.MenuBackground.Count;` is better ( you can combine both if statements).

**Review by lodicolo — CHANGES_REQUESTED** [review]

File: Intersect.Client/Core/Graphics.cs [hunk 17]
```diff
@@ -142,15 +149,80 @@ public static void DrawIntro()
...
+                {
+                    sMenuBackgroundIndex =
+                        (sMenuBackgroundIndex + 1) % ClientConfiguration.Instance.MenuBackground.Count;
+
+                    sMenuBackgroundInterval = currentTimeMs + ClientConfiguration.Instance.MenuBackgroundFrameInterval;
+                }
+            }
+
+            // Static background in the main menu.
+            else if (ClientConfiguration.Instance.MenuBackground.Count == 1)
```
Comment (lodicolo) [inline Intersect.Client/Core/Graphics.cs:184]: No longer needs to be `else if`, just `else` now.

**Review by lodicolo — CHANGES_REQUESTED** [review]

**Review by lodicolo — APPROVED** [review]
